### PR TITLE
(MODULES-6323) EmbeddedInstance support

### PR DESCRIPTION
**Pull request by jpogran**
This commit adds support for EmbeddedInstances (complex types and CIM Instances) when using `Invoke-DscResource` with the generic dsc type.

This commit also updates the `Usage` section of the README to reflect the usage of the dsc type. It does not attempt to update the entire README to reflect the changes from the dsc module, as that is out of scope for this PR.

### CODE REVIEW

File: lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb [hunk 1]
```diff
@@ -38,17 +37,41 @@ def self.format_boolean(value)
...
+          type       = value['dsc_type'].gsub('[]','')
+          properties = value['dsc_properties']
+          output = []
+          
+          properties = properties.is_a?(Hash) ? [properties] : properties
+          properties.each do |p|
+            output << "(New-CimInstance -ClassName '#{type}' -ClientOnly -Property #{format_hash_to_string(p)})"
+          end
+          output = output.join(",\n")
+          "[CimInstance[]]@(\n" + output +  "\n)"
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb]: I think this is missing the non-array variation of `[CimInstance]` usage.

See https://github.com/puppetlabs/puppetlabs-dsc/blob/master/lib/puppet/provider/templates/dsc/invoke_dsc_resource.ps1.erb#L45-L50

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb [hunk 2]
```diff
@@ -13,43 +13,130 @@
...
+        expected = <<-HERE
 @{
 'ensure' = 'Present';
-'bindinginfo' = @{
-'dsc_type' = 'MSFT_xWebBindingInformation[]';
-'dsc_properties' = @{
+'name' = 'foo';
+'state' = 'Started';
+'physicalpath' = 'c:/inetpub/wwwroot';
+'bindinginfo' = [CimInstance[]]@(
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: So for this scenario where the manifest has an EmbeddedInstance that is not a `String[]` (in the MOF schema), but just a single `String` instance like:

```
"dsc_type" => "MSFT_xWebBindingInformation",
"dsc_properties" => {}
```

I would expect that the PowerShell code does not use an array

```powershell
[CimInstance](New-CimInstance -ClassName 'MSFT_xWebBindingInformation' -ClientOnly -Property @{})
```

Which is how it works in DSC module at https://github.com/puppetlabs/puppetlabs-dsc/blob/master/lib/puppet/provider/templates/dsc/invoke_dsc_resource.ps1.erb#L45-L50

The original commit to DSC module has the info on why we did that (using MSFT_xWebsite as the example) - https://github.com/puppetlabs/puppetlabs-dsc/commit/cadf58fbc63c0193501b62f5b7e992a5969b825d

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb [hunk 3]
```diff
@@ -13,43 +13,130 @@
...
+      end
+
+      it "should output correct syntax with array CimInstance" do
+        expected = <<-HERE
+@{
+'ensure' = 'Present';
+'name' = 'foo';
+'state' = 'Started';
+'physicalpath' = 'c:/inetpub/wwwroot';
+'bindinginfo' = [CimInstance[]]@(
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: 👍

For `MSFT_xWebBindingInformation[]` this is how I expect the PowerShell code is rendered based on how DSC module works.

**Comment by Iristyle** [comment]
Just a couple of additional comments.

### On `[CimInstance]()` vs `[CimInstance]@()`

In your manifest example for the comment on the PR, you have `MSFT_xWebBindingInformation` in the manifest, and then pass an array in the manifest like `dsc_properties => []`.  The tests are a little different, so maybe that was an early example that changed?  

While we *could* make that syntax work, I think it's probably better to stick with something more explicit like `MSFT_xWebBindingInformation[]`.  While we can infer a `[CimInstance]@()` based on `dsc_properties => []`, we can't always determine if `dsc_properties => {}` is a single instance in an EmbeddedInstance array *or* if it's just a single instance, because it's ambiguous. If we get it wrong, the PowerShell call fails as described in https://github.com/puppetlabs/puppetlabs-dsc/commit/cadf58fbc63c0193501b62f5b7e992a5969b825d.

Alternatively, since it appears you've got `dsc_properties` working as either `[]` or `{}`, we could just use that to infer the right thing (requiring additionally that users always put the `[]` around single instanced arrays so we have a 1:1 mapping for generating the right code).

Any thoughts on which syntax is preferable?  If what I'm trying to convey is confusing, let me know.

### `MSFT_KeyValuePair`

I *think* we still need the special-handling for `MSFT_KeyValuePair`... used in spots like https://github.com/puppetlabs/puppetlabs-dsc/blob/3773d00b7f0cd475ee848bcaee0728d7bcfb03c1/lib/puppet_x/dsc_resources/xPSDesiredStateConfiguration/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.schema.mof#L8

In DSC module, that was added as part of https://github.com/puppetlabs/puppetlabs-dsc/commit/2816ce283f5943272df968ecedb2b746d843164a

Do we want to tackle that in another commit for this ticket or file a new ticket?  It doesn't look like it's used much in actual resources that are MOF based (but maybe more prevalent in class-based resources)?

File: README.md [hunk 4]
```diff
@@ -52,9 +48,9 @@ See [known issues](#known-issues) for troubleshooting setup.
 
 ## Usage
 
-### Using DSC Resources with Puppet
+### Using PSCredential or MSF_Credential
```
Comment (michaeltlombardi) [inline README.md]: Minor fix: MSF**T**_Credential.

File: README.md [hunk 5]
```diff
@@ -52,9 +48,9 @@ See [known issues](#known-issues) for troubleshooting setup.
...
 
-You can use a DSC Resource by prefixing each DSC Resource name and parameter with 'dsc_' and lowercasing the values.
+The generic `dsc` type is a streamlined and minimal representation of a DSC Resource declaration in Puppet syntax. You can use a DSC Resource by supplying the same properties you would set in a DSC Configuration script inside the `dsc_resource_properties` parameter. For most use cases the `dsc_resource_properties` parameter accepts the same structure as the PowerShell syntax, with the substituion of Puppet syntax for arrays, hashes and other data structures.
```
Comment (michaeltlombardi) [inline README.md]: May want to add here that you now need to specify the module and resource. It's clear in the example, I think, but could be spelled out in this paragraph.

File: README.md [hunk 6]
```diff
@@ -68,273 +64,133 @@ WindowsFeature IIS {
...
+        "wakka" => "woot",
+        "number"     => 8090
+      }
+    }
+  }
 }
 ~~~
 
-#### Registry Example: Dword and Qword
+### Using PSCredential or MSF_Credential
```
Comment (michaeltlombardi) [inline README.md]: Same nit as above.

File: README.md [hunk 7]
```diff
@@ -68,273 +64,133 @@ WindowsFeature IIS {
...
 }
 ~~~
 
-*Note*: DSC Resources are executed under the SYSTEM context by default, which means you are unable to access any user level Registry key without providing alternate credentials.
-
-### Adding or Removing Windows Features
+### Using CimInstance
 
-You can add or remove Windows Features using Puppet type `dsc_windowsfeature` which maps to the `WindowsFeature` DSC Resource.
+A DSC Resource may need a more complex type that a simple key value pair, so an EmbeddedInstance is used. An EmbeddedInstance are serialized as CimInstance over the wire. In order to represent a CimInstance in the `dsc` type, we will use the `dsc_type` key to specify which CimInstance to use. If the CimInstance is an array, we append a `[]` to the end of the name.
```
Comment (michaeltlombardi) [inline README.md]: > An embeddedInstance ~are~ **is a**...

File: README.md [hunk 8]
```diff
@@ -68,273 +64,133 @@ WindowsFeature IIS {
...
-
-~~~powershell
-[PS]> Get-WindowsFeature
-~~~
 
-### Website Installation Example
+### Handling Reboots with DSC
 
-An end-to-end example installation of a test website.
+Add the following `reboot` resource to your manifest. It must have the name `dsc_reboot` for the `dsc` module to find and use it.
```
Comment (michaeltlombardi) [inline README.md]: Nit: Can we clarify when this takes effect? I can kinda grok it from the example.

Comment (jpogran) [reply README.md]: Out of scope for this PR. This PR only has to do with implementing complex types and documenting them

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb (hunk 2 above)
Comment (jpogran) [reply spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: This was solved by relying on the user specifying '[]' in the `dsc_type` key.

**Comment by jpogran** [comment]
It was decided to rely on the user entering in (and keeping synced) two pieces of information when dealing with EmbeddedInstances that can either be a scalar or an array.

The `dsc_type` value must contain a `[]` if the value is an array. This was explained in the updates to the README

File: README.md (hunk 8 above)
Comment (Iristyle) [reply README.md]: Yeah, let's address the `reboot` situation later.  FWIW, the current solution doesn't work as intended anyhow (i.e. Puppet waits until *all* DSC resources have been evaluated before enacting the reboot, rather than triggering immediately after the DSC resource that requests it - see https://tickets.puppetlabs.com/browse/MODULES-3769).

File: lib/puppet/type/dsc.rb [hunk 9]
```diff
@@ -44,7 +44,11 @@
   end
 
   newparam(:dsc_resource_properties, :array_matching => :all) do
-    desc "DSC Resource Properties"
+    desc <<-HERE
+    The hash of properties to pass to the DSC Resource.
+
+    To express EmbeddedInstances, the `dsc_resource_properties` parameter will reconize any key with a hash value that contains two keys: `dsc_type` and `dsc_properties`, as a indication of how to format the data supplied. The `dsc_type` contains the CimInstance name to use, and the `dsc_properties` contains a hash or an array of hashes representing the data for the CimInstances. If the CimInstance is an array, we append a `[]` to the end of the name.
```
Comment (Iristyle) [inline lib/puppet/type/dsc.rb]: nit: `recognize`

Comment (Iristyle) [reply lib/puppet/type/dsc.rb]: Can we use backticks in inline docs?

File: lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb [hunk 10]
```diff
@@ -38,17 +37,47 @@ def self.format_boolean(value)
...
+          end
+        end
+
+        def self.format_ciminstance(value)
+          type       = value['dsc_type'].gsub('[]','')
+          properties = value['dsc_properties']
+          output     = []
+          properties = properties.is_a?(Hash) ? [properties] : properties
+
+          properties.each do |p|
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb]: Can remove `output = []` and shorten this to 

```ruby
output = properties.map do |p|
  "(New-CimInstance -ClassName '#{type}' -ClientOnly -Property #{format_hash_to_string(p)})"
end
```

File: lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb [hunk 11]
```diff
@@ -38,17 +37,47 @@ def self.format_boolean(value)
...
+              "([PSCustomObject]#{format_hash(value['dsc_properties'])} | new-pscredential)"
+            else
+              format_ciminstance(value)
+            end
+          end
+        end
+
+        def self.format_ciminstance(value)
+          type       = value['dsc_type'].gsub('[]','')
+          properties = value['dsc_properties']
```
Comment (Iristyle) [inline lib/puppet_x/puppetlabs/dsc_lite/powershell_hash_formatter.rb]: Probably not as clear, but a common Ruby trick to always get an array here is:

```ruby
properties = [value['dsc_properties']].flatten
```

i.e.

```
[15] pry(main)> [{}].flatten
=> [{}]
[16] pry(main)> [[{}, {}]].flatten
=> [{}, {}]
```

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb (hunk 2 above)
Comment (Iristyle) [reply spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: All these examples look good now.

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb [hunk 12]
```diff
@@ -13,44 +13,189 @@
...
+        })
+        expect(result).to eq expected.strip
+      end
+
+      it "should output correct syntax with MSFT_KeyValuePair" do
+        expected = <<-HERE
+@{
+'destinationPath' = 'c:\fileName.jpg';
+'uri' = 'http://www.contoso.com/image.jpg';
+'userAgent' = '[Microsoft.PowerShell.Commands.PSUserAgent]::InternetExplorer';
```
Comment (Iristyle) [inline spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: Was this intended to be the literal string `'[Microsoft.PowerShell.Commands.PSUserAgent]::InternetExplorer'` or was this intended to return the constant from the .NET class?

Comment (jpogran) [reply spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: It has to be quoted in the Puppet manifest, or it breaks the parser. Since it's a string in the Puppet manifest, it's parsed as a string and set as quoted in the PowerShell hash. Technically the PowerShell hash could take a type or a string as parsable, however it's irrelevant as PowerShell will cast whatever is passed to the property based on the parameter declaration of the DSC Resource. In this case, it's typed as a string and passed in a PSBoundParameters block directly to Invoke-WebRequest, which has UserAgent as a string parameter.

Even if we can't rely on the type casting PowerShell does under the covers, I don't see  this being an issue in the future, for if it needs to be a 'complex type' then it needs to be an EmbeddedInstance class declaration so it can be serialized into a CimInstance.

**Comment by jpogran** [comment]
@michaeltlombardi I've addressed some of your notes about the readme by adding more to the types section and fleshing out some of the builtin type desc blocks.

File: spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb (hunk 12 above)
Comment (Iristyle) [reply spec/integration/puppet_x/puppetlabs/powershell_hash_formatter_spec.rb]: Yeah, I understood the manifest -> generated code mechanics, but didn't connect the dots to the PS conversion.

So this ends up being a pretty nifty trick that we can rely on PS conversion for stringified value -> actual .NET type. 

This is good to know / something we should probably consider documenting for users.

https://blogs.msdn.microsoft.com/powershell/2013/06/11/understanding-powershells-type-conversion-magic/

File: README.md [hunk 13]
```diff
@@ -68,282 +65,162 @@ WindowsFeature IIS {
...
-  dsc_hash_parameter => {
-    'key1' => 'value1',
-    'key2' => 'value2'
-  },
-}
-~~~
+### Using CimInstance
 
-### Using Credentials
+A DSC Resource may need a more complex type that a simple key value pair, so an EmbeddedInstance is used. An EmbeddedInstance is serialized as CimInstance over the wire. In order to represent a CimInstance in the `dsc` type, we will use the `dsc_type` key to specify which CimInstance to use. If the CimInstance is an array, we append a `[]` to the end of the name.
```
Comment (Iristyle) [inline README.md]: minor - `may need a more complex type than`

File: README.md [hunk 14]
```diff
@@ -68,282 +65,162 @@ WindowsFeature IIS {
...
-### Handling Reboots with DSC
-
-Add the following `reboot` resource to your manifest. It must have the name `dsc_reboot` for the `dsc` module to find and use it.
+A contrived, but simple example follows:
 
 ~~~puppet
-reboot { 'dsc_reboot' :
-  message => 'DSC has requested a reboot',
-  when => 'pending'
+dsc{'foo':
```
Comment (Iristyle) [inline README.md]: Minor formatting - other examples use `dsc {`, not `dsc{`

File: README.md [hunk 15]
```diff
@@ -68,282 +65,162 @@ WindowsFeature IIS {
...
-
-    $record = $View.GetType().InvokeMember("Fetch", "InvokeMethod", $Null, $View, $Null)
-    $property = $record.GetType().InvokeMember("StringData", "GetProperty", $Null, $record, 1)
-
-    $View.GetType().InvokeMember("Close", "InvokeMethod", $Null, $View, $Null)
-
-    return $property
-  }catch{
-    throw "Failed to read MSI file. The error was: {0}." -f $_
+dsc{'foouser':
```
Comment (Iristyle) [inline README.md]: Minor formatting - other examples use `dsc {`, not `dsc{`

File: README.md [hunk 16]
```diff
@@ -68,282 +65,162 @@ WindowsFeature IIS {
...
-  dsc_username             => 'jane-doe',
-  dsc_description          => 'Jane Doe user',
-  dsc_ensure               => present,
-  dsc_password             => {
-    'user' => 'jane-doe',
-    'password' => 'jane-password'
-  },
-  dsc_passwordneverexpires => false,
-  dsc_disabled             => true,
+dsc{'NewWebsite':
```
Comment (Iristyle) [inline README.md]: Minor formatting - other examples use `dsc {`, not `dsc{`

File: README.md [hunk 17]
```diff
@@ -68,282 +65,162 @@ WindowsFeature IIS {
...
-
-You can use either `ensure =>` (Puppet's `ensure`) or `dsc_ensure =>` (DSC's `Ensure`) in your manifests for Puppet DSC resource types. If you use both in a Puppet DSC resource, `dsc_ensure` overrides the value in `ensure`, so the value for `ensure` is essentially ignored.
+For the simplest cases, the above example is enough. However there are more advanced use cases in DSC that require more custom syntax in the `dsc` Puppet type. Since the `dsc` Puppet type has no prior knowledge of the type for each property in a DSC Resource, it can't format the hash correctly without some hints.
 
-We recommend that you use `dsc_ensure` instead of `ensure`, as it is a closer match for converting the DSC properties to Puppet DSC resources. It also overrides `ensure`, so there is less confusion if both are accidentally included.
+The `dsc_resource_properties` parameter will recognize any key with a hash value that contains two keys: `dsc_type` and `dsc_properties`, as a indication of how to format the data supplied. The `dsc_type` contains the CimInstance name to use, and the `dsc_properties` contains a hash or an array of hashes representing the data for the CimInstances.
```
Comment (Iristyle) [inline README.md]: We should probably link out to / explain a bit more about what a `CimInstance` is... for a future update to README

File: lib/puppet/type/dsc.rb [hunk 18]
```diff
@@ -44,7 +44,11 @@
   end
 
   newparam(:dsc_resource_properties, :array_matching => :all) do
-    desc "DSC Resource Properties"
+    desc <<-HERE
+    The hash of properties to pass to the DSC Resource.
+
+    To express EmbeddedInstances, the dsc_resource_properties parameter will reconize any key with a hash value that contains two keys: dsc_type and dsc_properties, as a indication of how to format the data supplied. The dsc_type contains the CimInstance name to use, and the dsc_properties contains a hash or an array of hashes representing the data for the CimInstances. If the CimInstance is an array, we append a [] to the end of the name.
```
Comment (Iristyle) [inline lib/puppet/type/dsc.rb]: Still says 'reconize` instead of `recognize`

**Comment by Iristyle** [comment]
We can address the minor typos, etc in #19 I think?
